core/consensus/common/src/import_queue.rs

```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
 	/// Import bunch of blocks.
 	fn import_blocks(&self, origin: BlockOrigin, blocks: Vec<IncomingBlock<B>>);
 	/// Import a block justification.
-	fn import_justification(&self, hash: B::Hash, number: NumberFor<B>, justification: Justification) -> bool;
+	fn import_justification(&self, who: Origin, hash: B::Hash, number: NumberFor<B>, justification: Justification);
+}
+
+pub trait ImportQueueClone<B: BlockT> {
+	fn clone_box(&self) -> Box<ImportQueue<B>>;
+}
+
+impl<B: BlockT> Clone for Box<ImportQueue<B>> {
+	fn clone(&self) -> Box<ImportQueue<B>> {
+		self.clone_box()
+	}
 }
 
 /// Import queue status. It isn't completely accurate.
+#[derive(Debug)]
 pub struct ImportQueueStatus<B: BlockT> {
 	/// Number of blocks that are currently in the queue.
 	pub importing_count: usize,
 	/// The number of the best block that was ever in the queue since start/last failure.
 	pub best_importing_number: <<B as BlockT>::Header as HeaderT>::Number,
 }
 
-/// Basic block import queue that is importing blocks sequentially in a separate thread,
+/// Interface to a basic block import queue that is importing blocks sequentially in a separate thread,
 /// with pluggable verification.
-pub struct BasicQueue<B: BlockT, V: 'static + Verifier<B>> {
-	handle: Mutex<Option<::std::thread::JoinHandle<()>>>,
-	data: Arc<AsyncImportQueueData<B>>,
-	verifier: Arc<V>,
-	block_import: SharedBlockImport<B>,
-	justification_import: Option<SharedJustificationImport<B>>,
+#[derive(Clone)]
+pub struct BasicQueue<B: BlockT> {
+	sender: Sender<BlockImportMsg<B>>,
 }
 
-/// Locks order: queue, queue_blocks, best_importing_number
-pub struct AsyncImportQueueData<B: BlockT> {
-	signal: Condvar,
-	queue: Mutex<VecDeque<(BlockOrigin, Vec<IncomingBlock<B>>)>>,
-	queue_blocks: RwLock<HashSet<B::Hash>>,
-	best_importing_number: RwLock<<<B as BlockT>::Header as HeaderT>::Number>,
-	is_stopping: AtomicBool,
-}
-
-impl<B: BlockT, V: Verifier<B>> BasicQueue<B, V> {
-	/// Instantiate a new basic queue, with given verifier and justification import.
-	pub fn new(verifier: Arc<V>, block_import: SharedBlockImport<B>, justification_import: Option<SharedJustificationImport<B>>) -> Self {
-		Self {
-			handle: Mutex::new(None),
-			data: Arc::new(AsyncImportQueueData::new()),
-			verifier,
-			block_import,
-			justification_import,
-		}
+impl<B: BlockT> ImportQueueClone<B> for BasicQueue<B> {
+	fn clone_box(&self) -> Box<ImportQueue<B>> {
+		Box::new(self.clone())
 	}
 }
 
-impl<B: BlockT> AsyncImportQueueData<B> {
-	/// Instantiate a new async import queue data.
-	pub fn new() -> Self {
+impl<B: BlockT> BasicQueue<B> {
```

Comment:
Can you add a general description here about how the queue works? e.g. how messages are dispatched to a worker, and blocks are imported on a different worker.